### PR TITLE
Reset the terminal if an exception occurs while handling terminal input in the app runner

### PR DESCRIPTION
**Pull request by jason-simmons**

### CODE REVIEW

**Comment by jason-simmons** [comment]
@B3rn475

**Comment by Hixie** [comment]
Looking at the code, I would have hoped this would be in a place that's obviously paired with the call to setupTerminal. The code that calls that also calls registerSignalHandlers.

I don't really understand the code flow here. There's a number of other places that call resetTerminal, which I would have expected to all be in one place.

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 1]
```diff
@@ -646,6 +646,9 @@ abstract class ResidentRunner {
       final bool handled = await _commonTerminalInputHandler(command);
       if (!handled)
         await handleTerminalCommand(command);
+    } catch (error) {
+      _resetTerminal();
+      rethrow;
```
Comment (B3rn475) [inline packages/flutter_tools/lib/src/resident_runner.dart]: Who is going to catch this exception?
If we reset the terminal shouldn't we remove the listener too?

**Comment by jason-simmons** [comment]
processTerminalInput is called within a microtask that is scheduled in response to pending input.

If processTerminalInput throws, currently nobody is catching the exception and the tool process will exit.  This patch keeps that behavior but cleans up the terminal before exiting.

**Comment by jason-simmons** [comment]
Changed this to log the exception and then call _cleanUpAndExit.  This will centralize the terminal reset in the same place where signals are handled.

**Review by B3rn475 — APPROVED** [review]

**Pull request merged**